snuba/datasets/sessions.py

```diff
@@ -0,0 +1,270 @@
+            "seq": message["seq"],
+            "deleted": 0,
+            "org_id": message["org_id"],
+            "project_id": message["project_id"],
+            "retention_days": message["retention_days"],
+            "duration": duration,
+            "status": STATUS_MAPPING[message["status"]],
+            "errors": _collapse_uint16(message["errors"]) or 0,
+            "timestamp": _ensure_valid_date(
+                datetime.utcfromtimestamp(message["timestamp"])
+            ),
+            "started": _ensure_valid_date(
+                datetime.utcfromtimestamp(message["started"])
+            ),
+            "release": message.get("release"),
+            "environment": message.get("environment"),
+        }
+        return ProcessedMessage(action=action_type, data=[processed])
+
+
+class SessionDataset(TimeSeriesDataset):
+    def __init__(self) -> None:
+        all_columns = ColumnSet(
+            [
+                ("session_id", UUID()),
+                ("distinct_id", UUID()),
+                ("seq", UInt(64)),
+                ("org_id", UInt(64)),
+                ("project_id", UInt(64)),
+                ("retention_days", UInt(16)),
+                ("duration", UInt(32)),
+                ("status", UInt(8)),
+                ("errors", UInt(16)),
+                # note here that deleted is really only an indicator so we can
+                # figure out which values can be discarded.  Deleting of values
+                # does not update the materializations.
+                ("deleted", UInt(8)),
+                ("timestamp", DateTime()),
+                ("started", DateTime()),
+                ("release", LowCardinality(Nullable(String()))),
+                ("environment", LowCardinality(Nullable(String()))),
+            ]
+        )
+
+        write_schema = ReplacingMergeTreeSchema(
+            columns=all_columns,
+            local_table_name=WRITE_LOCAL_TABLE_NAME,
+            dist_table_name=WRITE_DIST_TABLE_NAME,
+            prewhere_candidates=["project_id", "org_id"],
+            order_by="(org_id, project_id, toDate(started), session_id, cityHash64(toString(session_id)))",
+            partition_by="(toMonday(timestamp))",
+            version_column="seq",
+            sample_expr="cityHash64(toString(session_id))",
+            settings={"index_granularity": 16384},
+        )
+
+        read_columns = ColumnSet(
+            [
+                ("org_id", UInt(64)),
+                ("project_id", UInt(64)),
+                ("started", DateTime()),
+                ("release", LowCardinality(Nullable(String()))),
+                ("environment", LowCardinality(Nullable(String()))),
+                (
+                    "duration",
+                    AggregateFunction("quantilesIf(0.5, 0.9)", UInt(32), UInt(8)),
+                ),
+                ("uniq_sessions", AggregateFunction("countIf", UUID(), UInt(8))),
```

Comment:
With the current ReplacingMergeTree in the raw table this has to be a uniqIf, if instead we have the init event this can be countIf, but then we cannot use a ReplacingMergeTree in the raw table otherwise the init event will be merged together with the crash/end/whatever event and disappear.
Also if this is not a uniqIf I would not call it uniq_sessions but count_sessions